DatadogCore/Sources/Core/Upload/DataUploadDelay.swift

```diff
@@ -9,12 +9,12 @@ import DatadogInternal
 
 internal protocol Delay {
     var current: TimeInterval { get }
-    mutating func decrease()
-    mutating func increase()
+    func decrease()
+    func increase()
 }
 
 /// Mutable interval used for periodic data uploads.
-internal struct DataUploadDelay: Delay {
+internal class DataUploadDelay: Delay {
```

Comment:
It's solving the problem of passing the copy. I think it was mainly causing issues in tests (inability to inspect the injected Delay). But technically it prevents from capturing the immutable copy of delay within closures (which I did in the very first version of the PR).

Reply:
I'm on a fence with this tbh.

having them mutating enforces the usage of var instead of let, but in reality mutability is achieved internally in Delay implementation.

Reply:
> having them mutating enforces the usage of var instead of let

I don't see what's wrong about that, especially with `private` accessor. `Delay` protocol has no implementation, and specifying `mutating` is required because you don't know if the provided implementation is a class or a struct.

```swift
protocol Delay {
    var current: TimeInterval { get } // <- This is expected to be mutated
    mutating func decrease()
    mutating func increase()
}
```
The other option is to constrain the `Delay` to classes with `protocol Delay: AnyObject` but I don't see the need for that.

Reply:
The only problem I see it that using mutating enforces `DataUploadWorker` to carry `var` delay which gives a false impression of mutability. In reality it's a class with some internal logic, which actually in some cases doesn't even mutate the state (for example when it reaches top/bottom value of the delay).

I think that `mutating` was added because initially implementation of the Delay was based on struct. Now mutability is achieved by altering the current value internally in `DataUploadDelay`. IMO from the protocol level we don't need to point out that mutability.

Reply:
I'm sorry to re-open this minor change, it's a nit-pick but I would like to keep things semantically correct..

`Delay` is a protocol, not a class nor a struct, the `mutating` is defined because the protocol could be applied on any type, on reference types or on value types. Its `current` property will be mutated by both of the methods, and the `DataUploadWorker` doesn't know what type of object the delay instance is, therefore it's a `var`. The previous implementation was correct, and safe with the `private` accessor.

Now if you really want to get rid of the `mutating` keyword, you have to constrain the `Delay` protocol to be class-only with:
```swift
protocol Delay: AnyObject {
   var current: TimeInterval { get }
   func decrease()
   func increase()
}
```
That's IMHO unnecessary, but it would be valid 👍

Reply:
As discussed protocol is actually not needed in this case. I'll remove it :)